pint/quantity.py

```diff
@@ -79,6 +78,262 @@ def wrapped(self, *args, **kwargs):
+    return output_wrap(np.where(condition, *args))
+
+@implements(np.linspace)
+def _linspace(start, stop, *args, **kwargs):
+    (start, stop), output_wrap = unwrap_and_wrap_consistent_units(start, stop)
+    return output_wrap(np.linspace(start, stop, *args, **kwargs))
+
+@implements(np.concatenate)
+def _concatenate(sequence, *args, **kwargs):
+    sequence, output_wrap = unwrap_and_wrap_consistent_units(*sequence)
+    return output_wrap(np.concatenate(sequence, *args, **kwargs))
+
+@implements(np.stack)
+def _stack(arrays, *args, **kwargs):
+    arrays, output_wrap = unwrap_and_wrap_consistent_units(*arrays)
+    return output_wrap(np.stack(arrays, *args, **kwargs))
+
+@implements(np.compress)
+def _compress(condition, a, *args, **kwargs):
+    (a,), output_wrap = unwrap_and_wrap_consistent_units(a)
+    return output_wrap(np.compress(condition, a, *args, **kwargs))
+
+@implements(np.append)
+def _append(arr, values, axis=None):
+    (arr, values), output_wrap = unwrap_and_wrap_consistent_units(arr, values)
+    return output_wrap(np.append(arr, values, axis))
+
+@implements(np.clip)
+def _clip(a, a_min, a_max, *args, **kwargs):
+    (a, a_min, a_max), output_wrap = unwrap_and_wrap_consistent_units(a, a_min, a_max)
+    return output_wrap(np.clip(a, a_min, a_max, *args, **kwargs))
+
+@implements(np.nan_to_num)
+def _nan_to_num(x, copy=True, nan=None, posinf=None, neginf=None):
+    if nan is None:
+        nan = 0.0
+    elif isinstance(nan, Quantity):
+        nan = nan.m_as(x.units)
+    posinf = posinf if not isinstance(posinf, Quantity) else posinf.m_as(x.units)
+    neginf = neginf if not isinstance(neginf, Quantity) else neginf.m_as(x.units)
+    result_magnitude = np.nan_to_num(x.magnitude, copy=False, nan=nan, posinf=posinf,
+                                     neginf=neginf)
+    if not copy:
+        x._magnitude = result_magnitude
+
+    return x._REGISTRY.Quantity(result_magnitude, x.units)
+
+@implements(np.isclose)
+def _isclose(a, b, rtol=None, atol=None, equal_nan=False):
+    (a, b, rtol, atol), _ = unwrap_and_wrap_consistent_units(a, b, rtol, atol)
+    rtol = 1e-05 if rtol is None else rtol
+    atol = 1e-08 if atol is None else atol
+    return np.isclose(a, b, rtol, atol, equal_nan)
+
+@implements(np.searchsorted)
+def _searchsorted(a, v, *args, **kwargs):
+    (a, v), _ = unwrap_and_wrap_consistent_units(a, v)
+    return np.searchsorted(a, v, *args, **kwargs)
+
+# Handle single unit argument operations (axis ops, aggregations, etc.)
+for func_str in ['expand_dims', 'squeeze', 'rollaxis', 'moveaxis', 'fix', 'around', 'diagonal',
+                 'mean', 'ptp', 'ravel', 'round_', 'sort', 'median', 'nanmedian', 'transpose',
+                 'flip', 'copy', 'trim_zeros', 'average', 'nanmean']:
+    func = getattr(np, func_str)
+    @implements(func)
+    def implementation(a, *args, **kwargs):
+        (a,), output_wrap = unwrap_and_wrap_consistent_units(a)
+        return output_wrap(func(a, *args, **kwargs))
```

Comment:
there's something wrong with this and all similar closures below: at the end of quantity, all functions are implemented by wrapping `atleast_3d` (look at `HANDLED_FUNCTIONS[func].__globals__["func"]`).

I'm no expert on closures so I don't know why this happens or how to solve this (I will look into it, though). Does anyone else have any suggestions?

Reply:
looks like creating `implementation` in a wrapper function (similar to `implement_func`) will solve this. We also should look at `HANDLED_FUNCTIONS[func].__closure__` and `HANDLED_FUNCTIONS[func].__code__.co_freevars` instead.